x/multi_coin_rewards/types/types.go

```diff
@@ -0,0 +1,51 @@
+package types
+
+import (
+	"fmt"
+
+	"cosmossdk.io/math"
+)
+
+type (
+	MultiCoinDistributionMap                 map[string][]MultiCoinDistributionPoolNormalizedEntry
+	MultiCoinDistributionPoolNormalizedEntry struct {
+		PoolId           uint64
+		NormalizedWeight math.LegacyDec
+	}
+)
+
+func ParseMultiCoinDistributionMap(policy MultiCoinDistributionPolicy) (MultiCoinDistributionMap, error) {
+	var distributionMap MultiCoinDistributionMap = make(map[string][]MultiCoinDistributionPoolNormalizedEntry)
+
+	for _, denomEntry := range policy.Entries {
+		distributionWeightsDuplicateCheck := make(map[uint64]struct{})
+
+		totalWeight := math.LegacyNewDec(0)
+		for _, weights := range denomEntry.PoolWeights {
+			if !weights.Weight.IsPositive() {
+				return nil, fmt.Errorf("invalid weight for pool id %d", weights.PoolId)
+			}
+			totalWeight = totalWeight.Add(weights.Weight)
+			if _, ok := distributionWeightsDuplicateCheck[weights.PoolId]; ok {
```

Comment:
this should be checked once in `SetMultiCoinRewardDistributionPolicy` instead of checking this on every 50th block again and again